lib/IDETool/IDEInspectionInstance.cpp

```diff
@@ -331,6 +334,17 @@ bool IDEInspectionInstance::performCachedOperationIfPossible(
                            1);
     SM.setIDEInspectionTarget(newBufferID, newOffset);
 
+    MapSourceLocIntoOriginalBuffer = [&SM, startOffset, tmpBufferID,
+                                      newBufferID](SourceLoc Loc) -> SourceLoc {
+      if (SM.findBufferContainingLoc(Loc) == newBufferID) {
+        return SM.getLocForOffset(tmpBufferID,
+                                  SM.getLocOffsetInBuffer(Loc, newBufferID) +
+                                      startOffset);
+      } else {
+        return Loc;
+      }
+    };
+
```

Comment:
AFAICT, `tmpBufferID` is a buffer ID in `tmpSM`, not `SM`. So I believe `SM.getLocForOffset(tmpBufferID, ...)` is not correct, and this PR currently happens to work because the bufferID is the same `1`.

Also, I think this whole `SourceLoc` conversion mechanism should be handled in `SourceManager`.
I think we need to:
a) get offset, line, and column in the original buffer
b) compare source loc between beyond buffer IDs.

For (a) line and columns can be retrieved by `getPresumedLineAndColumnForLoc`. For `offset`, I would add an optional parameter to `SourceManager::openVirtualFile()` like `offsetOffset` (I'm serious) and remember it in `VirtualFiles`, then create `SourceManager::getPresumedLocOffsetInBuffer()` so retrieve the adjusted "offset".

For (b), `SourceManager` knows the replaced source ranges, in `ReplacedRanges`, so it should be able to "compare" the locations beyond source buffer if the loc is a part of `ReplacedRanges` (like `ASTScope` does).

Could you explore this option? I feel passing around this `MapSourceLocIntoOriginalBuffer` is a too much burden.